### PR TITLE
Changing get_element_by_attribute() and get_element_by_class() to return a list of results. See PR #11956

### PR DESCRIPTION
**Pull request by ThomasChr**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/rg3/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/rg3/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/rg3/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [x] Improvement
- [ ] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

Explanation of your *pull request* in arbitrary form goes here. Please make sure the description explains the purpose and effect of your *pull request* and is worded well enough to be understood. Provide as much context and examples as possible.

As proposed by @yan12125 

> This is my idea:

    get_element_by_id is not touched (by HTML standards IDs are unique)
    get_element_by_attribute => get_elements_by_attribute
    get_element_by_class => get_elements_by_class

The two functions get_element_by_attribute and get_element_by_class are now returning a list of results.
The have also been renamed and all extractors changed to use the new functions.

### CODE REVIEW

**Comment by dstftw** [comment]
Old functions must be preserved untouched.

**Comment by ThomasChr** [comment]
@dstftw That was exactly what I proposed to @yan12125 in PR #11956 .
And what he didn't want! See the PR #11956 and see the message at top of this PR.

Could you please agree on that before I make any changes to this PR?
Changing all Extractors to use the new functions was a lot of work for me which could have been avoided...

**Comment by dstftw** [comment]
Rename of any util method breaks all 3rdparty code that uses it. This is unwanted and should be avoided.

**Comment by ThomasChr** [comment]
@dstftw Thats my opinion also. So I just add the new functions in this PR!

**Comment by yan12125** [comment]
@dstftw For me only the ```YoutubeDL``` class should have stable APIs. But if you really want to preserve old function names, I suggest to call new functions from old functions, mark old functions as deprecated and use new functions in all extractors. The deprecation can be documented in change logs. After a few months deprecated functions can be removed.

**Comment by dstftw** [comment]
>For me only the ```YoutubeDL``` class should have stable APIs.

I see no difference between `YoutubeDL` and `utils`. Both are encouraged to be used in code. Both should keep the interfaces.

>mark old functions as deprecated and use new functions in all extractors

Why deprecate at all? They should stay at least for convenience. If one only wants to get the first attribute with `get_elements_*` he'll be forced to do `[0]` all the time that is tedious and adds unnecessary clutter.

**Comment by ThomasChr** [comment]
So everyone fine with this version?
I just added the new functions without using them anywhere.

I would like to get this changes to close PR #11956 and Issue #11953

**Comment by dstftw** [comment]
1. Code duplication in single and multiple function versions should be eliminated.
2. Commits should be squashed.

**Comment by ThomasChr** [comment]
@dstftw 
1. If you mind telling me how you would like to achieve that? Calling the old function in a loop won't work without a ugly loop which cuts of parts of the html. I'm open for suggestions.
2. I can squash them, no problem - I only thought you simpley squash them when applying the PR.

**Comment by yan12125** [comment]
> They should stay at least for convenience

I believe codes with get_element_by_class is problematic - it conflicts with how HTML and Javascript's DOM API work. But nothing is impossible, so I have no objections on keeping them.

@ThomasChr Calling new functions from the original ones. And don't forget to update tests for new functions.

**Comment by ThomasChr** [comment]
@yan12125 Yes, calling the new functions from the old ones is the essiest way. Should have seen that.

So I'm:
- Calling the new functions from the old ones
- Write Tests for the new functions
- Squash all my commits to one

Anything else?

**Comment by dstftw** [comment]
@yan12125 how does *give-me-the-first-element-found-with-this-classname* scenario conflicts with DOM?

**Comment by yan12125** [comment]
In Javascript there's getElementsByClassName, getElementById but no getElementByClassName. To get the first element with the given class name, web developers have to use ```getElementsByClassName('foobar')[0]```. If there's really such a use case, ```get_elements_by_class_name('foobar')[0]``` is slightly better than ```get_element_by_class_name('foobar')```, IMO. Basically the concept of "first element with the given class" is not a primitive in DOM manipulations.

**Comment by dstftw** [comment]
>Basically the concept of "first element with the given class" is not a primitive in DOM manipulations.

Well, who cares? youtube-dl is not a web application. It does not even treat HTML as DOM in `get_element_by_attribute` rather it just dumb parse it with regex. So I don't see any point rigorously following concepts that semantically correct in DOM here. We aim to provide brief, concise and convenient set of APIs for data extraction to be used in conjunction youtube-dl, not to strictly follow some 	idiomatics inherent to some field X.

**Comment by yan12125** [comment]
@dstftw Please read my previous responses carefully. My latest attitude is not rejecting preserving get_element_by_class. I just dislike it.

**Comment by ThomasChr** [comment]
With this commit:
- The new functions are included
- The old functions call the new ones
- I have made three more tests
- All tests work fine

Everyone satisfied now?

File: youtube_dl/utils.py [hunk 1]
```diff
@@ -337,34 +337,53 @@ def get_element_by_id(id, html):
...
-    res = m.group('content')
+        if res.startswith('"') or res.startswith("'"):
+            res = res[1:-1]
+
+        retlist.append(unescapeHTML(res))
 
-    if res.startswith('"') or res.startswith("'"):
-        res = res[1:-1]
+    if not retlist:
+        return None
```
Comment (dstftw) [inline youtube_dl/utils.py]: It should always return a list.

File: youtube_dl/utils.py [hunk 2]
```diff
@@ -337,34 +337,53 @@ def get_element_by_id(id, html):
...
+    else:
+        return None
+
+
+def get_element_by_attribute(attribute, value, html, escape_value=True):
+    retval = get_elements_by_attribute(attribute, value, html, escape_value)
+    if retval:
+        return retval[0]
+    else:
+        return None
```
Comment (dstftw) [inline youtube_dl/utils.py]: This can be simplified to `return retval[0] if retval else None`.

File: youtube_dl/utils.py [hunk 3]
```diff
@@ -337,34 +337,53 @@ def get_element_by_id(id, html):
...
+
+def get_elements_by_class(class_name, html):
+    return get_elements_by_attribute(
         'class', r'[^\'"]*\b%s\b[^\'"]*' % re.escape(class_name),
         html, escape_value=False)
 
 
-def get_element_by_attribute(attribute, value, html, escape_value=True):
+def get_elements_by_attribute(attribute, value, html, escape_value=True):
     """Return the content of the tag with the specified attribute in the passed HTML document"""
```
Comment (dstftw) [inline youtube_dl/utils.py]: This doc string does not match the function now.

File: youtube_dl/utils.py [hunk 4]
```diff
@@ -337,34 +337,53 @@ def get_element_by_id(id, html):
...
 def get_element_by_class(class_name, html):
-    return get_element_by_attribute(
+    retval = get_elements_by_class(class_name, html)
+    if retval:
+        return retval[0]
+    else:
+        return None
+
+
+def get_element_by_attribute(attribute, value, html, escape_value=True):
```
Comment (dstftw) [inline youtube_dl/utils.py]: It worth adding a doc string.

File: youtube_dl/utils.py (hunk 2 above)
Comment (ThomasChr) [inline youtube_dl/utils.py]: Wow. Did not know the syntax. Thanks!

File: youtube_dl/utils.py (hunk 3 above)
Comment (ThomasChr) [inline youtube_dl/utils.py]: """Return the content of the first tag with the specified ID in the passed HTML document""" - okay?

Comment (dstftw) [inline youtube_dl/utils.py]: It's a list of elements not the first element.

**Comment by ThomasChr** [comment]
Changed everything and melted into the first commit. That's okay for you?

**Comment by dstftw** [comment]
Check code with flake8.

**Comment by ThomasChr** [comment]
Checked with flake8, found one line with 4 trailing whitespaces :+1:

**Comment by dstftw** [comment]
`.\test\test_utils.py:1137:87: W291 trailing whitespace`

**Comment by ThomasChr** [comment]
Thats so embarrassing. I did not check the utils file :-(
Checked, corrected and recommited.

**Comment by ThomasChr** [comment]
@dstftw Thanks for approving. I'm learning quite a lot and I'm glad to help!

**Comment by ThomasChr** [comment]
Now that it's approved, could anyone remove the "pending fixes" label?

**Pull request merged**